models/issue_indexer.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/blevesearch/bleve/search/query"
+	"strings"
+)
+
+// indexerUpdateQueue queue of issues that need to be updated in the issues
+// indexer
+var indexerUpdateQueue chan *Issue
+
+// IssueIndexerData issue data stored in the issues indexer
+type IssueIndexerData struct {
+	ID int64
+	RepoID int64
+
+	Title string
+	Content string
+}
+
+// numericQuery an numeric-equality query for the given value and field
+func numericQuery(value int64, field string) *query.NumericRangeQuery {
+	f := float64(value)
+	tru := true
+	q := bleve.NewNumericRangeInclusiveQuery(&f, &f, &tru, &tru)
+	q.SetField(field)
+	return q
+}
+
+// SearchIssuesByKeyword searches for issues by given conditions.
+// Returns the matching issue IDs
+func SearchIssuesByKeyword(repoID int64, keyword string) ([]int64, error) {
+	fields := strings.Fields(strings.ToLower(keyword))
+	indexerQuery := bleve.NewConjunctionQuery(
+		numericQuery(repoID, "RepoID"),
+		bleve.NewDisjunctionQuery(
+			bleve.NewPhraseQuery(fields, "Title"),
+			bleve.NewPhraseQuery(fields, "Content"),
+		))
+	search := bleve.NewSearchRequestOptions(indexerQuery, 2147483647, 0, false)
+	search.Fields = []string{"ID"}
+
+	index, err := bleve.Open(setting.IssueIndexerPath)
+	if err != nil {
+		return nil, err
+	}
+	result, err := index.Search(search)
+	if err != nil {
+		index.Close()
+		return nil, err
+	}
+	if err = index.Close(); err != nil {
+		return nil, err
+	}
+
+	issueIDs := make([]int64, len(result.Hits))
+	for i, hit := range result.Hits {
+		issueIDs[i] = int64(hit.Fields["ID"].(float64))
+	}
+	return issueIDs, nil
+}
+
+// InitIssuesIndexer initialize issue indexer
+func InitIssuesIndexer() error {
+	_, err := os.Stat(setting.IssueIndexerPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = CreateIssuesIndexer(); err != nil {
+				return err
+			}
+			go func() {
+				if err := PopulateIssuesIndexer(); err != nil {
+					log.Error(4, "PopulateIssuesIndex: %v", err)
+				}
+			}()
+		} else {
+			return err
+		}
+	}
+	indexerUpdateQueue = make(chan *Issue, 10)
+	go processIndexerUpdateQueue()
+	return nil
+}
+
+// CreateIssuesIndexer create an issues indexer
+func CreateIssuesIndexer() error {
+	mapping := bleve.NewIndexMapping()
+	docMapping := bleve.NewDocumentMapping()
+
+	docMapping.AddFieldMappingsAt("ID", bleve.NewNumericFieldMapping())
+	docMapping.AddFieldMappingsAt("RepoID", bleve.NewNumericFieldMapping())
+
+	textFieldMapping := bleve.NewTextFieldMapping()
+	textFieldMapping.Analyzer = simple.Name
+	docMapping.AddFieldMappingsAt("Title", textFieldMapping)
+	docMapping.AddFieldMappingsAt("Content", textFieldMapping)
+
+	mapping.AddDocumentMapping("issues", docMapping)
+	index, err := bleve.New(setting.IssueIndexerPath, mapping)
+	if err != nil {
+		return err
+	}
+	return index.Close()
+}
+
+// PopulateIssuesIndexer write all issues in the database to the issues indexer
+func PopulateIssuesIndexer() error {
+	index, err := bleve.Open(setting.IssueIndexerPath)
```

Comment:
`index.Close()` will always be called; every return statement is preceded by a call to `index.Close()`. I decided not to use `defer index.Close()` since `index.Close()` returns an `error`, and I wanted to be able to check if closing the index was successful.

If you want, I can change it to use `defer index.Close()`